docs/auth.md

```diff
@@ -0,0 +1,176 @@
+      window.localStorage.setItem(key, entries[key]);
+    });
+  }
+}, [localStorage]);
+```
+
+### Session storage
+Session storage ([`window.sessionStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage))
+is specific to a particular domain.
+
+```js
+// Get session storage and store as env variable
+const sessionStorage = await page.evaluate(() => JSON.stringify(sessionStorage));
+process.env.SESSION_STORAGE = sessionStorage;
+
+// Set session storage in a new context
+const sessionStorage = process.env.SESSION_STORAGE;
+await context.addInitScript(([storage]) => {
+  if (window.location.hostname === 'example.com') {
+    const entries = JSON.parse(storage);
+    Object.keys(entries).forEach(key => {
+      window.sessionStorage.setItem(key, entries[key]);
+    });
+  }
+}, [sessionStorage]);
+```
+
+### Lifecycle
+
+1. Run tests (for example, with `npm run test`)
+2. Login via UI and retrieve authentication state
+    * In Jest, this can be executed in [`globalSetup`](https://jestjs.io/docs/en/configuration#globalsetup-string).
+3. In each test, load authentication state in `beforeEach` or `beforeAll` step.
+
+This lifecycle can be extended to cache authentication state across test runs.
+In that case, the expiry of authentication state (cookies/tokens in storage) would
+need to be handled.
+
+This approach to reuse authentication state would **work in CI environments**, since
+it does not rely on any external state.
+
+### Example
+
+[This script](examples/authentication.js) logs in on GitHub.com with Chromium,
+and then reuses the logged in cookie state in WebKit.
+
+### API reference
+- [class `BrowserContext`](./api.md#class-browsercontext)
+- [`browserContext.cookies`](./api.md#browsercontextcookiesurls)
+- [`browserContext.addCookies`](./api.md#browsercontextaddcookiescookies)
+- [`page.evaluate`](./api.md#pageevaluatepagefunction-arg)
+- [`browserContext.addInitScript`](./api.md#browsercontextaddinitscriptscript-arg)
+
+## Persistent authentication
+Web browsers use a directory on disk to store user history, cookies and other local
+state. This is directory is called the [User data directory](https://chromium.googlesource.com/chromium/src/+/master/docs/user_data_dir.md).
+
+User data directory is an alternate approach to reusing authentication state. It
+relies on a persistent location on disk, and therefore is not suited for CI
+environments. User data directories are browser-specific, and cannot be shared
+across browser types.
+
+The `launchPersistentContext` API enables user data directories.
+
+```js
+const { chromium } = require('playwright');
+
+const userDataDir = '/path/to/directory';
+const context = await chromium.launchPersistentContext(userDataDir);
+// Execute login steps
+```
+
+#### API reference
+- [class `BrowserContext`](./api.md#class-browsercontext)
+- [`browserType.launchPersistentContext`](./api.md#browsertypelaunchpersistentcontextuserdatadir-options)
+
+## Multi-factor authentication
```

Comment:
This sounds like a better sell for persistent context. Let's merge with previous section?